pkg/assetfetcher/http.go

```diff
@@ -58,57 +57,62 @@ type NexposeAssetFetcher struct {
 // FetchAssets gets all the assets for a given site ID from Nexpose, with pagination.
 // It returns a channel of assets and a channel of errors that can by asynchronously listened to.
 func (c *NexposeAssetFetcher) FetchAssets(ctx context.Context, siteID string) (<-chan domain.Asset, <-chan error) {
-	errChan := make(chan error)
-	assetChan := make(chan domain.Asset)
-	go func() {
-		var wg sync.WaitGroup
-		defer close(assetChan)
-		defer close(errChan)
-		// make the first call to Nexpose to get the total number of pages we'll
-		// have to page through to get all the assets, start with page 0
-		var currentPage = 0
-		req, err := newNexposeSiteAssetsRequest(c.Host, siteID, currentPage, c.PageSize)
-		if err != nil {
-			errChan <- err
-			return
-		}
-
-		res, err := http.DefaultClient.Do(req.WithContext(ctx))
-		if err != nil {
-			errChan <- &NexposeHTTPRequestError{err, req.URL.String()}
-			return
-		}
-		defer res.Body.Close()
-		respBody, err := ioutil.ReadAll(res.Body)
-		if err != nil {
-			errChan <- &ErrorReadingNexposeResponse{err, req.URL.String()}
-			return
-		}
-
-		var siteAssetResp SiteAssetsResponse
-		if err := json.Unmarshal(respBody, &siteAssetResp); err != nil {
-			errChan <- &ErrorParsingJSONResponse{err, req.URL.String()}
-			return
-		}
-
-		for _, asset := range siteAssetResp.Resources {
-			assetChan <- asset
-		}
-
-		// We've gotten the first page (page 0) and added the assets to the channel,
-		// so here we'll increment the currentPage to start on page 1 and use TotalPages, which is
-		// the total number of pages of assets that Nexpose has, to paginate through to get all the assets
+	errChan := make(chan error, 1)
+	defer close(errChan)
+
+	// have to page through to get all the assets, start with page 0
+	// make the first call to Nexpose to get the total number of pages we'll
+	var currentPage = 0
+	req, err := newNexposeSiteAssetsRequest(c.Host, siteID, currentPage, c.PageSize)
+	if err != nil {
+		errChan <- err
+		return nil, errChan
+	}
+
+	res, err := http.DefaultClient.Do(req.WithContext(ctx))
```

Comment:
:no_mouth: I realize now that this is easily the interpretation of what I said earlier. What I meant about using the `http.DefaultClient` was that the handler should still accept the client as a parameter but we should inject `http.DefaultClient` in main. This will solve for two problems. One is that it will enable you to inject a mock `http.RoundTripper` for the tests rather than needing a server somewhere that can receive an network call. The second is that it will stage us later to inject a smarter client when we start putting code in AWS Lambda where we won't be able to leverage our proxy.

The suggestion of removing the smart client setup from the code and using the default client was meant to help standardize this repo with the others we're making. The goal will be to make a sweeping pass and add optional smart client support to all the projects in the same way.

Reply:
Welp, that's unfortunate. I quite liked ripping out all the code associated with passing around the http client everywhere. Is this a blocker? I can see how it wold help with writing the tests, but those are already written now. Do you think it will be worth making this change and changing all the tests? Or just recreating the httpClient in main.go and passing it to the handler?

Reply:
I'd be OK with deferring so long as it gets a tracking ticket and is changed before we ship. If we shipped this way then this would be the only service that works this way and it will create a one-off set of work to refactor in the future to bring it back into alignment when we actually start bringing in smart clients.

From a testing perspective, they may be already written but these tests are also a one-off compared to all of our other tests for this kind of code. They require orchestration of an external dependency (the test server) and rely on having a network available for they to succeed. These facts break from our standard model for unit testing which avoids all external dependencies and orchestration through mocking.

Reply:
I've added the creation and passing in of the HTTP client back to main.go and I'm working on updating the tests.